pkg/kv/kvserver/replica_proposal_buf.go

```diff
@@ -874,19 +875,26 @@ func proposeBatch(
 		p.onErrProposalDropped(ents, props, raftGroup.BasicStatus().RaftState)
 		return nil //nolint:returnerrcheck
 	}
-	if err == nil {
-		// Now that we know what raft log position[1] this proposal is to end up
-		// in, deduct flow tokens for it. This is done without blocking (we've
-		// already waited for available flow tokens pre-evaluation). The tokens
-		// will later be returned once we're informed of the entry being
-		// admitted below raft.
-		//
-		// [1]: We're relying on an undocumented side effect of upstream raft
-		//      API where it populates the index and term for the passed in
-		//      slice of entries. See etcd-io/raft#57.
-		maybeDeductFlowTokens(ctx, p.flowControlHandle(ctx), handles, ents)
+	if err != nil {
+		return err
 	}
-	return err
+	// Now that we know what raft log position[1] this proposal is to end up
+	// in, deduct flow tokens for it. This is done without blocking (we've
+	// already waited for available flow tokens pre-evaluation). The tokens
+	// will later be returned once we're informed of the entry being
+	// admitted below raft.
+	//
+	// [1]: We're relying on an undocumented side effect of upstream raft
+	//      API where it populates the index and term for the passed in
+	//      slice of entries. See etcd-io/raft#57.
+	maybeDeductFlowTokens(ctx, p.flowControlHandle(ctx), handles, ents)
+
+	// Register the proposal with rafttrace. This will add the trace to the raft
+	// lifecycle. We trace at most one entry per batch, so pick the last one as
+	// is the last one to be completed.
+	lastIndex := len(ents) - 1
+	p.registerForTracing(props[lastIndex], ents[lastIndex])
```

Comment:
The last entry does not necessarily contain a span. One entry can contain one, another entry can not contain one. These are independent across entries, and might correspond to different requests/transactions etc. I think we need a loop here.

Reply:
I changed back to a loop but register at most once per call. This is hopefully a reasonable tradeoff, but we can see if it causes issues with missed traces and re-evaulate.